src/Formats/ReadSchemaUtils.cpp

```diff
@@ -236,4 +239,37 @@ NamesAndTypesList getNamesAndRecursivelyNullableTypes(const Block & header)
     return result;
 }
 
+String getKeyForSchemaCache(const String & source, const String & format, const std::optional<FormatSettings> & format_settings, const ContextPtr & context)
+{
+    return getKeysForSchemaCache({source}, format, format_settings, context).front();
+}
+
+static String makeSchemaCacheKey(const String & source, const String & format, const String & additional_format_info)
+{
+    return source + "@@" + format + "@@" + additional_format_info;
+}
```

Comment:
TODO: replace to struct (later)

Reply:
https://github.com/ClickHouse/ClickHouse/pull/40414